crates/ark/src/lsp/inputs/library.rs

```diff
@@ -0,0 +1,148 @@
+//
+// library.rs
+//
+// Copyright (C) 2025 by Posit Software, PBC
+//
+
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::sync::RwLock;
+
+use super::package::Package;
+use crate::lsp;
+
+/// Lazily manages a list of known R packages by name
+#[derive(Default, Clone, Debug)]
+pub struct Library {
+    /// Paths to library directories, i.e. what `base::libPaths()` returns.
+    pub library_paths: Arc<Vec<PathBuf>>,
+
+    packages: Arc<RwLock<HashMap<String, Option<Arc<Package>>>>>,
+}
+
+impl Library {
+    pub fn new(library_paths: Vec<PathBuf>) -> Self {
+        Self {
+            packages: Arc::new(RwLock::new(HashMap::new())),
+            library_paths: Arc::new(library_paths),
+        }
+    }
+
+    /// Get a package by name, loading and caching it if necessary.
+    /// Returns `None` if the package can't be found or loaded.
+    pub fn get(&self, name: &str) -> Option<Arc<Package>> {
+        // Try to get from cache first (could be `None` if we already tried to
+        // load a non-existent or broken package)
+        if let Some(entry) = self.packages.read().unwrap().get(name) {
+            return entry.clone();
+        }
+
+        // Not cached, try to load
+        let pkg = match self.load_package(name) {
+            Ok(Some(pkg)) => Some(Arc::new(pkg)),
+            Ok(None) => None,
+            Err(err) => {
+                lsp::log_error!("Can't load R package: {err:?}");
+                None
+            },
+        };
+
+        self.packages
+            .write()
+            .unwrap()
+            .insert(name.to_string(), pkg.clone());
+
+        pkg
+    }
+
+    /// Insert a package in the library for testing purposes.
+    #[cfg(test)]
+    pub fn insert(self, name: &str, package: Package) -> Self {
+        self.packages
+            .write()
+            .unwrap()
+            .insert(name.to_string(), Some(Arc::new(package)));
+        self
+    }
+
+    fn load_package(&self, name: &str) -> anyhow::Result<Option<Package>> {
+        for lib_path in self.library_paths.iter() {
+            match Package::load(&lib_path, name) {
+                Ok(Some(pkg)) => return Ok(Some(pkg)),
+                Ok(None) => (),
+                Err(err) => lsp::log_warn!("Can't load package: {err:?}"),
+            }
```

Comment:
In the `Err` case, we should probably bail with the error right?

Right now I think this says "I found the package, but failed to load it"

And IMO we should _not_ continue on to the next `library_paths` element when that occurs

Reply:
If you decide not to do that, note that you don't need the return value to be `anyhow::Result<>` here, as you never propagate an error right now